engine.lisp

```diff
@@ -148,9 +168,69 @@ Returns T if all constraints match, NIL otherwise."
   (loop for constraint in (cdr matcher)
         always (match-constraint constraint node)))
 
+(declaim (ftype (function (character list plump:node (function (plump:node) T))
+                          (values &optional null))
+                match-pair-depth))
+(defun match-pair-depth (combinator matcher parent extender)
+  "Match a combinator and matcher pair against a list of nodes.
+Returns a vector of matching nodes."
```

Comment:
This docstring does not seem to be valid given that it does not in fact return anything and instead uses an extender. Furthermore, extender here seems to be a misnomer, given that the function is not only used to accumulate a value, but also to recursively process it in a group. Naming it something like `matching-nodes-processor` should be more adequate.